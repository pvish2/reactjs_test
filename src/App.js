// src/App.js

import React, { Component } from "react";
import Constants from "./constants";
import Card from "./components/cards";

class App extends Component {
  constructor(props) {
    super(props);
    this.state = {
      cards: [],
      errorVisible: false,
      message: ""
    };
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  componentDidMount() {
    this.getAllCreditCardList();
  }

  getAllCreditCardList() {
    fetch(Constants.API_URL)
      .then(res => res.json())
      .then(data => {
        this.setState({ cards: data });
      })
      .catch(e => console.log(e));
  }

  handleSubmit(event) {
    event.preventDefault();
    let data = {};
    data["name"] = document.getElementById("name").value;
    data["cardNumber"] = document.getElementById("cardNumber").value;
    data["limit"] = document.getElementById("limit").value;

    if (!this.checkValidation(data)) {
      return;
    }
    fetch(Constants.API_URL, {
      headers: new Headers({ "content-type": "application/json" }),
      method: "POST",
      body: JSON.stringify(data)
    })
      .then(res => res.json())
      .then(data => {
        if (data.status === 1) {
          this.setState({ errorVisible: false, message: data.message });
          this.getAllCreditCardList();
        } else {
          this.setState({ errorVisible: true, message: data.message });
        }
      })
      .catch(e => console.log(e));
  }

  checkValidation(data) {
    if (data.cardNumber.length !== 10) {
      this.setState({ errorVisible: true, message: Constants.INVALID_CARD });
      return false;
    }
    if (data.limit < 0) {
      this.setState({
        errorVisible: true,
        message: "limit should be more than 0"
      });
      return false;
    }
    return true;
  }

  render() {
    return (
      <div class="parentDiv">
        <h2>Credit Card System</h2>
        <h4 class="addDiv"> Add</h4>
        <div>
          <form id="form1" onSubmit={this.handleSubmit}>
            <div class="form-group">
              <label for="name">Name</label>
              <input type="text" class="form-control" id="name" required />
            </div>
            <div class="form-group">
              <label for="cardNumber">Card Number</label>
              <input
                type="number"
                class="form-control"
                id="cardNumber"
                required
              />
            </div>
            <div class="form-group">
              <label for="limit">Limit</label>
              <input type="number" class="form-control" id="limit" required />
            </div>
            <button type="submit" class="btn btn-dark">
              Add
            </button>
          </form>
        </div>
        {this.state.errorVisible ? (
          <h4 class="errorDiv"> {this.state.message}</h4>
        ) : null}

        <div></div>
        <h4 class="existingDiv">Existing Cards</h4>

        <Card cards={this.state.cards} />
      </div>
    );
  }
}

export default App;
