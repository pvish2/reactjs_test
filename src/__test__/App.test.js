import React from "react";
import { shallow } from "enzyme";
import App from "../App.js";
import Constants from "../constants";

describe("App", () => {
  let component = "";

  beforeAll(() => {
    component = shallow(<App />);
  });

  it("should render my app", () => {
    expect(component.getElements()).toMatchSnapshot();
  });

  it("fetches data from server when server returns a successful response", done => {
    const mockSuccessResponse = {};
    const mockJsonPromise = Promise.resolve(mockSuccessResponse);
    const mockFetchPromise = Promise.resolve({
      json: () => mockJsonPromise
    });
    jest.spyOn(global, "fetch").mockImplementation(() => mockFetchPromise);
    shallow(<App />);
    expect(global.fetch).toHaveBeenCalledTimes(1);
    expect(global.fetch).toHaveBeenCalledWith(Constants.API_URL);

    process.nextTick(() => {
      expect(component.state()).toEqual({
        cards: [],
        errorVisible: false,
        message: ""
      });
      global.fetch.mockClear();
      done();
    });
  });

  it("should show error on invalid card", () => {
    const mockData = {
      name: "test",
      cardNumber: "765757577",
      limit: 12
    };
    component.instance().checkValidation(mockData);
    expect(component.state().errorVisible).toEqual(true);
    expect(component.state().message).toEqual(Constants.INVALID_CARD);
  });

  it("should show error on invalid limit", () => {
    const mockData = {
      name: "test",
      cardNumber: "0437093859",
      limit: -12
    };
    component.instance().checkValidation(mockData);
    expect(component.state().errorVisible).toEqual(true);
    expect(component.state().message).toEqual(Constants.INVALID_LIMIT);
  });
});
