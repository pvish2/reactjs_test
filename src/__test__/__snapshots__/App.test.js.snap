// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should render my app 1`] = `
Array [
  <div
    class="parentDiv"
  >
    <h2>
      Credit Card System
    </h2>
    <h4
      class="addDiv"
    >
       Add
    </h4>
    <div>
      <form
        id="form1"
        onSubmit={[Function]}
      >
        <div
          class="form-group"
        >
          <label
            for="name"
          >
            Name
          </label>
          <input
            class="form-control"
            id="name"
            required={true}
            type="text"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="cardNumber"
          >
            Card Number
          </label>
          <input
            class="form-control"
            id="cardNumber"
            required={true}
            type="number"
          />
        </div>
        <div
          class="form-group"
        >
          <label
            for="limit"
          >
            Limit
          </label>
          <input
            class="form-control"
            id="limit"
            required={true}
            type="number"
          />
        </div>
        <button
          class="btn btn-dark"
          type="submit"
        >
          Add
        </button>
      </form>
    </div>
    <div />
    <h4
      class="existingDiv"
    >
      Existing Cards
    </h4>
    <Cards
      cards={Array []}
    />
  </div>,
]
`;
